tests/e2e/Services/Projects/ProjectsDevKeys.php

```diff
@@ -0,0 +1,265 @@
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $id,
+            'x-appwrite-development-key' => $response['body']['secret']
+        ], [
+            'email' => 'user@appwrite.io',
+            'password' => 'password'
+        ]);
+        $this->assertEquals('429', $res['headers']['status-code']);
+    }
+
+
+    /**
+     * @depends testCreateProjectDevKey
+     * @group devKeys
+     */
+    public function testUpdateProjectDevKey($data): array
+    {
+        $id = $data['projectId'] ?? '';
+        $keyId = $data['keyId'] ?? '';
+
+        $response = $this->client->call(Client::METHOD_PUT, '/projects/' . $id . '/development-keys/' . $keyId, array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), [
+            'name' => 'Key Test Update',
+            'expire' => DateTime::addSeconds(new \DateTime(), 360),
+        ]);
+
+        $this->assertEquals(200, $response['headers']['status-code']);
+        $this->assertNotEmpty($response['body']['$id']);
+        $this->assertEquals($keyId, $response['body']['$id']);
+        $this->assertEquals('Key Test Update', $response['body']['name']);
+        $this->assertArrayHasKey('sdks', $response['body']);
+        $this->assertEmpty($response['body']['sdks']);
+        $this->assertArrayHasKey('accessedAt', $response['body']);
+        $this->assertEmpty($response['body']['accessedAt']);
+
+        $response = $this->client->call(Client::METHOD_GET, '/projects/' . $id . '/development-keys/' . $keyId, array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), []);
+
+        $this->assertEquals(200, $response['headers']['status-code']);
+        $this->assertNotEmpty($response['body']['$id']);
+        $this->assertEquals($keyId, $response['body']['$id']);
+        $this->assertEquals('Key Test Update', $response['body']['name']);
+        $this->assertArrayHasKey('sdks', $response['body']);
+        $this->assertEmpty($response['body']['sdks']);
+        $this->assertArrayHasKey('accessedAt', $response['body']);
+        $this->assertEmpty($response['body']['accessedAt']);
+
+        return $data;
+    }
+
+    /**
+     * @depends testCreateProjectDevKey
+     * @group devKeys
+     */
+    public function testDeleteProjectDevKey($data): array
+    {
+        $id = $data['projectId'] ?? '';
+        $keyId = $data['keyId'] ?? '';
+
+        $response = $this->client->call(Client::METHOD_DELETE, '/projects/' . $id . '/development-keys/' . $keyId, array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), []);
+
+        $this->assertEquals(204, $response['headers']['status-code']);
```

Comment:
After deleting, i would try to use it, and ensure behaviour there is expected (exception, or rate limits)

Reply:
updated